NBitcoin.Altcoins/Stratis.cs

```diff
@@ -0,0 +1,660 @@
+				nNonce = 0;
+			}
+
+			private byte[] CalculateHash(byte[] data, int offset, int count)
+			{
+				byte[] hashData = data.Skip(offset).Take(count).ToArray();
+				uint256 hash = null;
+				if (this.nVersion > 6)
+					hash = Hashes.Hash256(hashData);
+				else
+				{
+					var x13 = new X13();
+					hash = new uint256(this.x13.ComputeBytes(hashData));
+				}
+				return hash.ToBytes();
+			}
+
+			protected override HashStreamBase CreateHashStream()
+			{
+				return BufferedHashStream.CreateFrom(CalculateHash);
+			}
+
+			public override uint256 GetPoWHash()
+			{
+				
+				return new uint256(this.x13.ComputeBytes(this.ToBytes()));
+			}
+		}
+
+		/// <summary>
+		/// A POS block that contains the additional block signature serialization.
+		/// </summary>
+		public class StratisBlock : Block
+		{
+			/// <summary>
+			/// A block signature - signed by one of the coin base txout[N]'s owner.
+			/// </summary>
+			private StratisBlockSignature blockSignature = new StratisBlockSignature();
+
+			public StratisBlock(StratisBlockHeader blockHeader) : base(blockHeader)
+			{
+			}
+
+			public override ConsensusFactory GetConsensusFactory()
+			{
+				return StratisConsensusFactory.Instance;
+			}
+
+			/// <summary>
+			/// The block signature type.
+			/// </summary>
+			public StratisBlockSignature BlockSignature
+			{
+				get { return this.blockSignature; }
+				set { this.blockSignature = value; }
+			}
+
+			/// <summary>
+			/// The additional serialization of the block POS block.
+			/// </summary>
+			public override void ReadWrite(BitcoinStream stream)
+			{
+				base.ReadWrite(stream);
+				stream.ReadWrite(ref this.blockSignature);
+			}
+		}
+		
+		private class StratisWitness
+		{
+			private TxInList _Inputs;
+
+			public StratisWitness(TxInList inputs)
+			{
+				_Inputs = inputs;
+			}
+
+			internal bool IsNull()
+			{
+				return _Inputs.All(i => i.WitScript.PushCount == 0);
+			}
+
+			internal void ReadWrite(BitcoinStream stream)
+			{
+				for (int i = 0; i < _Inputs.Count; i++)
+				{
+					if (stream.Serializing)
+					{
+						var bytes = (_Inputs[i].WitScript ?? WitScript.Empty).ToBytes();
+						stream.ReadWrite(ref bytes);
+					}
+					else
+					{
+						_Inputs[i].WitScript = WitScript.Load(stream);
+					}
+				}
+
+				if (IsNull())
+					throw new FormatException("Superfluous witness record");
+			}
+		}
+
+		public class StratisTransaction : Transaction
+		{
+			public StratisTransaction(ConsensusFactory consensusFactory)
+			{
+				_Factory = consensusFactory;
+			}
+
+			ConsensusFactory _Factory;
+			public override ConsensusFactory GetConsensusFactory()
+			{
+				return _Factory;
+			}
+
+			/// <summary>
+			/// POS Timestamp
+			/// </summary>
+			public uint Time { get; set; } = Utils.DateTimeToUnixTime(DateTime.UtcNow);
+
+			public override void ReadWrite(BitcoinStream stream)
+			{		
+				var witSupported = (((uint)stream.TransactionOptions & (uint)TransactionOptions.Witness) != 0) &&
```

Comment:
This is okay of course but it looks to c++. My suggestion is:

```suggestion
				var witSupported = stream.TransactionOptions.HasFlag(TransactionOptions.Witness) &&
```